### PR TITLE
Feature/http sink

### PR DESCRIPTION
**Pull request by colmsnowplow**
Draft PR for http sink.

Notes

- It's not perfect but it works - likely a lot of improvements could be made.
- Current implementation omits qs for post, omits body for get, and sends a hardcoded get for HEAD
- We still generate data for all formats every time, which is inefficient. I can see how to refactor it to improve this but don't have the time - it's already in an issue.
- I think it'd be beneficial to be able to configure paths/weights of the different path options - I hope to get to adding this separately for this release, but we don't need it for collector so I'm first going to move to the deployment of this asset as an RC.
- There is a bug whereby the data generated itself is producing corrupt context data sometimes - which when base64-encoded cx field results in a JSON with the character '7' at the end (When decoded). It results in enrichment failures - so not critical to current use case but I'll log an issue.

TODO this release:

- [x] Address review feedback
- [x] Make distribution of GET/POST/HEAD configurable
- [x] Fix origin headers
- [ ] Fix HEAD requests (or do something sensible with them) - Won't do - the server gives an error. Just leaving unimplemented for now.

To make tickets for:

- [x] Bug whereby cx field in querystring produces a `7` character at the end of the b64 when decoded: https://github.com/snowplow-incubator/snowplow-event-generator/issues/60
- [x] Make paths distribution configurable - this interacts with request type so needs some config design consideration: https://github.com/snowplow-incubator/snowplow-event-generator/issues/61
- [x] Make body/querystring generation configurable - sometimes we want a specific combination of both so needs some config design consideration: https://github.com/snowplow-incubator/snowplow-event-generator/issues/62
- [x] Refactor things function so that we run only the Gen functions we need rather than running all of them: https://github.com/snowplow-incubator/snowplow-event-generator/issues/63
- [x] Refactor things such that the same Body data is used for all 3 gen functions - so that for the same seed, different formats of the same data are produced: https://github.com/snowplow-incubator/snowplow-event-generator/issues/64

### CODE REVIEW

File: sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala
Comment (istreeter) [inline sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala:17]: Unusual to rename from Upper case name to lower case name.  More normal to write `{Method => TrackerMethod}`

File: sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala
Comment (istreeter) [inline sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala:77]: A few comments about this little section:

Easy one -- consider using `IllegalArgumentException` instead of `Exception`.

Another easy one -- It's unusual to use the `.format` syntax in scala.  It's more normal to write it like

```
s"blah blah ${properties.endpoint} blah blah ${generatedRequest.method.path}"
```

A nice change you could make is to move the `Uri.fromString(...)` outside of this `buildRequest` block.  So you only run it once when the app first starts up.  And you only need to handle the error scenario once.

Then, once you have generated a `Uri`, you can use syntax like:

```
val address = baseUri / generatedRequest.method.path.vendor / generatedRequest.method.path.version
```

File: sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala
Comment (istreeter) [inline sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala:125]: Here when you call `httpClient.use { ..... }`  it creates and destroys the http client + all of the internal objects needed to implement it.  Like a TCP connection pool etc.

The problem is, you call `.use` for every single request.  So you're continually creating/destroying the resources, which is quite wasteful.

It's better to find a way to `.use` the http client just once, and then re-use it for every request.  I'd be happy to show you a few ways to arrange the code for this.

File: sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala
Comment (istreeter) [inline sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala:73]: I agree this is the right approach.  I think we can make it neater though!

File: sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Config.scala
Comment (istreeter) [inline sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Config.scala:72]: If you made endpoint a `org.http4s.Uri` instead of a `String` in the config, then you wouldn't need to handle parsing exceptions later on when generating requests.

**Comment by colmsnowplow** [comment]
Note: HEAD requests are still not implemented. Using `client.status` instead of `client.expect` does prevent the app from crashing with these requests, but they all get:

> This hits org.http4s.client.UnexpectedStatus: unexpected HTTP status: 422 Unprocessable Entity for request HEAD

Since I suspect that this is something to do with the request itself, rather than an expected error from the collector, I've left it for another day.

I have also made a [separate PR](https://github.com/snowplow-incubator/snowplow-event-generator/pull/65/files) to make request paths configurable. With the current implementation, we can generate `/r/*` paths, for which the collector correctly responds with 4XX or 3XX responses. Since it's desirable that for some of our tests (eg load stress testing) expect 100% of the data to succeed (ie. where failure responses are indicators of the limits of the infrastructure), and it was fairly simple to implement (while other work was blocked), I made the PR.

**Comment by colmsnowplow** [comment]
@istreeter it doesn't need your urgent attention, I've also flagged it to @peel who will likely soon be using the tool - but as an FYI/because I think you might be interested, see last comment for where things lie.

**Comment by colmsnowplow** [comment]
Note: Formatting might be off. I realised that I hadn't run `scalafmt` long after I had separately branched off. I plan to do that before release, just holding off until after I have to rebase one or more of these branches. :)

File: sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala
Comment (peel) [inline sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala:129]: Again, not a reason for changing anything. But maybe that's a personal preference, but I've always found a function should have a single reason to change. Never been a massive fan of internal methods especially as it makes them hard to test individually.

I'd structure this function in a way that address, body, parseHeaders are separate functions, this way buildRequest could potentially look like this:

```
def buildRequesst(generatedRequest: HttpRequest): Request[F] = generatedRequest.method match {
  case TrackerMethod.Post(_) =>  Request[F](...)
  ...
}

def address(_) = _
def body(_) = _ 
def parseHeaders(_) = _
```

**Review by peel — APPROVED** [review]
The idea of being able to tune HTTP request formats is fantastic and overall this feature is massively helpful. Great job!

Please do run formatting and remove dead code. A few extra (hopefully helpful) comments below.

File: sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala
Comment (colmsnowplow) [reply sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala:52]: Yeah so I didn't touch the data generation part, but I did feel that tweaks to that implementation could make our lives easier here/make our approach more idiomatic - which I think is what you're getting at here, if I'm understanding.

**Comment by colmsnowplow** [comment]
@peel / @istreeter I think it's ready for final review now, removed dead code and now parsing endpoint as Uri in config.

I will rebase commits before merging just didn't wanna make the convos hard to track. :)

**Pull request merged**
